src/Gemini/Modules/UndoRedo/Services/UndoRedoManager.cs

```diff
@@ -59,26 +81,56 @@ public void UndoAll()
 
         public void Redo(int actionCount)
         {
-            for (int i = 0; i < actionCount; i++)
+            OnBegin();
+
+            try
+            {
+                for (int i = 0; i < actionCount; i++)
+                {
+                    var action = Pop(_redoStack);
+                    action.Execute();
+                    Push(_undoStack, action);
+                }
+            }
+            finally
             {
-                var action = Pop(_redoStack);
-                action.Execute();
-                Push(_undoStack, action);
+                OnEnd();
             }
         }
 
         public void RedoTo(IUndoableAction action)
         {
-            while (true)
+            OnBegin();
+
+            try
+            {
+                while (true)
+                {
+                    var thisAction = Pop(_redoStack);
+                    thisAction.Execute();
+                    Push(_undoStack, thisAction);
+                    if (thisAction == action)
+                        return;
+                }
+            }
+            finally
             {
-                var thisAction = Pop(_redoStack);
-                thisAction.Execute();
-                Push(_undoStack, thisAction);
-                if (thisAction == action)
-                    return;
+                OnEnd();
             }
         }
 
+        private void OnBegin()
+        {
+            if (BatchBegin != null)
```

Comment:
Please could you change these to use the thread-safe style of raising an event? i.e.

``` csharp
var handler = BatchBegin;
if (handler != null)
    handler(this, EventArgs.Empty);
```